submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,312 @@
+will be closer to the Vanilla Submariner experience.
+
+## Design Details
+
+The following CRD is proposed to support the new design.
+
+```Go
+type ClusterGlobalEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of ClusterGlobalEgressIP object.
+    Spec               ClusterEgressIPSpec `json:"spec"`
+
+    // Observed status of NamespaceGlobalEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type ClusterEgressIPSpec struct {
+    // The number of globalIP's requested. Globalnet Controller will allocate the requested
+    // number of contiguous GlobalIPs for this ClusterGlobalEgressIP object.
+    // If unspecified, NumGlobalIPs defaults to 1.
+    // +optional
+    NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from all Namespaces.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+type NamespaceGlobalEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of NamespaceGlobalEgressIP object.
+    Spec               NamespaceEgressIPSpec `json:"spec"`
+
+    // Observed status of NamespaceGlobalEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type NamespaceEgressIPSpec struct {
+    // The number of globalIP's requested. Globalnet Controller will allocate the requested
+    // number of contiguous GlobalIPs for this NamespaceGlobalEgressIP object.
+    // If unspecified, NumGlobalIPs defaults to 1.
+    // +optional
+    NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from the namespace in which the
+    // CR is created.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+const (
+    // GlobalEgressIPSuccess means that Globalnet was able to successfully allocate the GlobalIP.
+    GlobalEgressIPSuccess GlobalEgressIPStatus = "Success"
+
+    // GlobalEgressIPError means that Globalnet was unable to allocate the GlobalIP.
+    GlobalEgressIPError GlobalEgressIPStatus = "Error"
+)
+
+type EgressIPStatus struct {
+    // Status is one of {"Success", "Error"}
+    Status GlobalEgressIPStatus `json:"status,omitempty"`
+
+    // +optional
+    Message *string `json:"message,omitempty"`
+
+    // When more than a single GlobalIP is allocated, the starting GlobalIP allocated to this object.
```

Comment:
I think a list of IPs is fine it's not an object that will be subject to changes, so the storage & API related traffic should not be significant, even if we had thousands of IPs. For example 2048 IPv4 would result in +32kB of data.

I would think differently if this type of object could be changing a lot, and subject to informers, etc.

Also, having the flexibility to provide non-contiguous IP would avoid fragmentation issues.

Reply:
As stated, IPs must be contiguous blocks so this design fits the limitations.

What is the added benefit of storing a list?

Reply:
I think we also have to restrict the number of GlobalIPs that can be requested. So added a max limit of 10.
In most of the cases, it will could be a single GlobalIP or maybe few GlobalIPs, so a list might make it easy to parse for the users. @mkolesnik I hope you are okay to have it as a list.

Regarding contiguous or non-contiguous, since the allocation will now be done by Globalnet (and not user-requested), lets leave it to implementation. IPTables based Globalnet requires them to be Contiguous. OVN based Globalnet might be able to support non-contiguous (not sure).